qiskit/circuit/library/standard_gates/u.py

```diff
@@ -273,33 +273,25 @@ def __array__(self, dtype=None):
 
     @property
     def params(self):
-        """Get parameters from base_gate.
-
-        Returns:
-            list: List of gate parameters.
-
-        Raises:
-            CircuitError: Controlled gate does not define a base gate
-        """
-        if self.base_gate:
-            # CU has one additional parameter to the U base gate
-            return self.base_gate.params + self._params
-        else:
-            raise CircuitError("Controlled gate does not define base gate for extracting params")
+        """return CU params."""
+        return self._params
 
     @params.setter
     def params(self, parameters):
-        """Set base gate parameters.
-
-        Args:
-            parameters (list): The list of parameters to set.
-
-        Raises:
-            CircuitError: If controlled gate does not define a base gate.
-        """
-        # CU has one additional parameter to the U base gate
-        self._params = [parameters[-1]]
+        """Set CUGate params [theta,phi,lam,gamma] and base_gate(UGate) params [theta,phi,lam]"""
+        self._params = []
+        for single_param in parameters:
+            if isinstance(single_param, ParameterExpression):
+                self._params.append(single_param)
+            else:
+                self._params.append(self.validate_parameter(single_param))
         if self.base_gate:
             self.base_gate.params = parameters[:-1]
         else:
             raise CircuitError("Controlled gate does not define base gate for extracting params")
+
+    def __deepcopy__(self, _memo=None):
+        """Include the parameters in the copy"""
+        cpy = super().__deepcopy__(_memo=_memo)
+        cpy.params = self.params.copy()
+        return cpy
```

Comment:
I'm guessing the reason this is necessary is because the parameter assignment isn't actually writing out to the `base_gate`, because `QuantumCircuit._assign_parameter` modifies the `Gate.params` object by setting an item in it.  That dodges the `params` setter, and since `params` isn't the same object as `base_gate.params` (which `ControlledGate` implicitly expects), we get the problem.

That said, because Qiskit makes such heavy use of `deepcopy` (because our ownership model of gates isn't great), _in practice_ this covers most of the times you might actually see that `base_gate` doesn't get updated.  For example, `QuantumCircuit.__eq__` implicitly deep copies both circuits because the equality is done via `DAGCircuit` (to resolve basic data-flow graph isomorphisms), and the conversion involves a copy.  You can still see that there's some issues, though:
```python
In [1]: from qiskit.circuit import Parameter, QuantumCircuit
   ...: a, b, c, d = (Parameter(x) for x in "abcd")
   ...:
   ...: qc1 = QuantumCircuit(2)
   ...: qc1.cu(0.1, 0.2, 0.3, 0.4, 0, 1)
   ...:
   ...: qc2 = QuantumCircuit(2)
   ...: qc2.cu(a, b, c, d, 0, 1)
   ...: qc2.assign_parameters([0.1, 0.2, 0.3, 0.4], inplace=True)
   ...: qc1 == qc2  # because there's a secret deepcopy
Out[1]: True

In [2]: qc1.data[0] == qc2.data[0]  # ... but _really_
Out[2]: False

In [3]: qc2.data[0].operation.params
Out[3]:
[ParameterExpression(0.1),
 ParameterExpression(0.2),
 ParameterExpression(0.3),
 ParameterExpression(0.4)]

In [4]: qc2.data[0].operation.base_gate.params
Out[4]: [Parameter(a), Parameter(b), Parameter(c)]
```

Reply:
I think some of the complexity may stem from the `@setter` for params being at the list level.   `QuantumCircuit._assign_parameter` retrieves a reference to  `Instruction._params` and  modifies a single element directly.  This process was causing the assignment to `CUGate.params` to be ineffective since the line `return self.base_gate.params + self._params` creates a new instance and the assignment is to this new copy (resulting in a no-op to the gate itself).

I included  `CUGate.__deepcopy__` since `CUGate._params` were no longer inherited from `Instruction` but managed at the `CUGate` level.   Without that overload any `QuantumCircuit.assign_parameters` that was not inline was creating two instances of the `CUGate` in two different circuits that shared a single `_params` instance. The equivalence_library copies were also sharing a single list instance with the original gate as well.

An alternative was to add a setter method for `_params` that operated at the individual element level instead of the list level.   I think this requires adding a `ParameterList()` class to manage the list.   That seems like the first step in refactoring how parameters are managed; since next it seems logical to move the assign/bind logic out of `QuantumCircuit` and into that new class. But then the relationship between the `ParameterList`  `QuantumCircuit` and `Instruction` didn't seem obvious to me.   Not sure what the appetite for that level of refactoring is, but I can open a issue for that if it is something worth considering.